arwen/contexts/runtime.go

```diff
@@ -74,56 +74,134 @@ func (context *runtimeContext) InitState() {
+
+	return false
+}
+
+func (context *runtimeContext) makeInstanceFromCompiledCode(codeHash []byte, gasLimit uint64, newCode bool) bool {
+	if !context.host.IsAheadOfTimeCompileEnabled() {
+		return false
+	}
+
+	if newCode || len(codeHash) == 0 {
+		return false
+	}
+
+	blockchain := context.host.Blockchain()
+	found, compiledCode := blockchain.GetCompiledCode(codeHash)
+	if !found {
+		return false
+	}
+
+	gasSchedule := context.host.Metering().GasSchedule()
+	options := wasmer.CompilationOptions{
+		GasLimit:           gasLimit,
+		UnmeteredLocals:    uint64(gasSchedule.WASMOpcodeCost.LocalsUnmetered),
+		OpcodeTrace:        false,
+		Metering:           true,
+		RuntimeBreakpoints: true,
+	}
+	newInstance, err := wasmer.NewInstanceFromCompiledCodeWithOptions(compiledCode, options)
+	if err != nil {
+		log.Trace("NewInstanceFromCompiledCodeWithOptions", "error", err)
+		return false
+	}
+
+	context.instance = newInstance
+
+	idContext := arwen.AddHostContext(context.host)
+	context.instance.SetContextData(idContext)
+
+	return true
+}
+
+func (context *runtimeContext) StartWasmerInstance(contract []byte, gasLimit uint64, newCode bool) error {
+	if context.RunningInstancesCount() >= context.maxWasmerInstances {
+		context.instance = nil
+		return arwen.ErrMaxInstancesReached
+	}
+
+	blockchain := context.host.Blockchain()
+	codeHash := blockchain.GetCodeHash(context.GetSCAddress())
+
+	warmInstanceSet := context.setWarmInstanceWhenNeeded()
+	if warmInstanceSet {
+		return nil
+	}
+
+	compiledCodeSet := context.makeInstanceFromCompiledCode(codeHash, gasLimit, newCode)
+	if compiledCodeSet {
+		return nil
+	}
+
+	log.Trace("Creating a new Wasmer instance")
+
+	gasSchedule := context.host.Metering().GasSchedule()
+	options := wasmer.CompilationOptions{
+		GasLimit:           gasLimit,
+		UnmeteredLocals:    uint64(gasSchedule.WASMOpcodeCost.LocalsUnmetered),
+		OpcodeTrace:        false,
+		Metering:           true,
+		RuntimeBreakpoints: true,
+	}
+	newInstance, err := wasmer.NewInstanceWithOptions(contract, options)
+	if err != nil {
+		context.instance = nil
+		return err
+	}
+
+	context.instance = newInstance
+	if context.useWarmInstance {
+		context.warmInstanceAddress = context.GetSCAddress()
+		context.warmInstance = context.instance
+	}
+
+	if newCode || len(codeHash) == 0 {
+		codeHash, err = context.host.Crypto().Sha256(contract)
 		if err != nil {
-			context.instance = nil
+			context.CleanWasmerInstance()
 			return err
 		}
+	}
 
-		context.instance = newInstance
-		if context.useWarmInstance {
-			context.warmInstanceAddress = context.GetSCAddress()
-			context.warmInstance = context.instance
-		}
+	context.saveCompiledCode(codeHash)
 
-		idContext := arwen.AddHostContext(context.host)
-		context.instance.SetContextData(idContext)
+	idContext := arwen.AddHostContext(context.host)
+	context.instance.SetContextData(idContext)
 
+	if newCode {
 		err = context.VerifyContractCode()
 		if err != nil {
 			context.CleanWasmerInstance()
```

Comment:
Also need to set `warmInstance` and `warmInstanceAddress` to `nil` here.

Reply:
done